### PR TITLE
[Hashing] Update code to use C++17

### PR DESCRIPTION
**Pull request by AZero13**
We can use std::apply to unpack the tuple arg and pass its elements as arguments to the lambda function.

### CODE REVIEW

**Comment by al45tair** [comment]
@swift-ci Please test

File: include/swift/AST/RequestCache.h [hunk 1]
```diff
@@ -29,10 +29,6 @@ namespace evaluator {
 
 namespace detail {
 
-// Remove this when the compiler bumps to C++17.
-template <typename...> using void_t = void;
-template<typename T, typename = void_t<>>
-
```
Comment (al45tair) [inline include/swift/AST/RequestCache.h:35]: Personally I have a preference for headers including the correct headers for things they use, so this one should probably now `#include <type_traits>`; I note, though that it's already missing `<tuple>` and probably other things, so it's not unlikely it currently builds OK.

**Review by al45tair — APPROVED** [review]
LGTM, though I think the title of the PR should probably be "[Hashing][Threading][AST] Update to C++17" or something along those lines, since there are lots of changes besides the `Hashing.h` one.

**Comment by RSilicon** [comment]
@al45tair Let's try that again

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by RSilicon** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by AreaZR** [comment]
@al45tair I believe this is ready now!

**Comment by al45tair** [comment]
@swift-ci Please test

**Pull request closed without merging**